### ISSUE TITLE
Parse Error on MODE message with more than one channel user modes for a single user

### ISSUE THREAD

**Issue by TorCapybara** [cinchrb/cinch#191]
This is a problem with more modern IRC Demons which try to be explicit about all modes a user has. I know +a is non standard, but it should at least parse +o correctly and not fail with a parse error. The MODE message is correctly formatted from the server as seen below.

The offending message of the server tries to give the Bot +a and +o in one MODE message, the MODE parser fails.
~~~
[2015/02/26 15:21:07.906] >> :ChanServ!ChanServ@services.* MODE #channel +ao CinchBot CinchBot 
~~~

The shortened stacktrace, it looks like the duplicate user gets removed so finally the parameter for the +o is missing and fails:
~~~
[2015/02/26 15:21:07.906] !! /home/user/.rvm/gems/ruby-2.1.2/bundler/gems/cinch-ba8d9f1c5a72/lib/cinch/mode_parser.rb:53:in `parse_modes': Too many parameters: +ao ["CinchBot"] (Cinch::Exceptions::InvalidModeString)
[2015/02/26 15:21:07.906] !!    /home/user/.rvm/gems/ruby-2.1.2/bundler/gems/cinch-ba8d9f1c5a72/lib/cinch/irc.rb:451:in `on_mode'
[2015/02/26 15:21:07.906] !!    /home/user/.rvm/gems/ruby-2.1.2/bundler/gems/cinch-ba8d9f1c5a72/lib/cinch/irc.rb:263:in `parse'
~~~

**Comment by dominikh** [comment]
Can you paste the 004 and all of the 005 messages? (They're sent when connecting). Alternatively, if it's a public network, tell me the address of the network.

**Comment by bartoszkopinski** [comment]
@dominikh These are my logs:
```
>> :irc.tinyspeck.com 004 irc.tinyspeck.com IRC-SLACK gateway
>> :irc.tinyspeck.com 001 hipbot Welcome to Slack IRC Gateway server
>> :irc.tinyspeck.com 005 hipbot MAP SILENCE=15 WHOX WALLCHOPS WALLVOICES USERIP CPRIVMSG CNOTICE MODES=6 MAXCHANNELS=100 SAFELIST :are supported by this server
>> :irc.tinyspeck.com 005 hipbot NICKLEN=32 TOPICLEN=160 AWAYLEN=160 CHANTYPES=# PREFIX=(ov)@+ CHANMODES=b,k,l,rimnpst CASEMAPPING=rfc1459 :are supported by this server
```
and the error:
```
>> :hipbot!hipbot@irc.tinyspeck.com JOIN #room
>> :irc.tinyspeck.com 332 hipbot #room :subject
<< WHO #room %acfhnru
>> :irc.tinyspeck.com 353 hipbot = #room :b hipbot hipbot-testing-user
<< MODE #room +b
>> :irc.tinyspeck.com 366 hipbot #room :End of /NAMES list.
<< MODE #room
>> :b!b@irc.tinyspeck.com MODE #room +v b : active
(Cinch::Exceptions::InvalidModeString)
!!  cinch-2.2.5/lib/cinch/irc.rb:451:in `on_mode'
!!  cinch-2.2.5/lib/cinch/irc.rb:263:in `parse'
!!  cinch-2.2.5/lib/cinch/irc.rb:151:in `block (2 levels) in start_reading_thread'
!!  cinch-2.2.5/lib/cinch/helpers.rb:123:in `rescue_exception'
!!  cinch-2.2.5/lib/cinch/irc.rb:149:in `block in start_reading_thread'
```
Any thoughts?

**Comment by dominikh** [comment]
@bartoszkopinski This is a separate issue. Slack is sending an invalid mode string (the `: active` part shouldn't be there.) – I'll see if I can add a workaround to Cinch to handle that.

**Comment by dominikh** [comment]
@TorCapybara I am still waiting on feedback from you.

**Comment by dominikh** [comment]
@bartoszkopinski I fixed the issue in 3bcd1b9 – Please test with that version and see if it works for you. It's possible (quite likely, really), that more Slack-specific errors will show up. Their implementation of "IRC" has very little to do with actual IRC.

**Comment by bartoszkopinski** [comment]
@dominikh thank you, same goes for their XMPP implementation. :( I'll try it out.

**Comment by dominikh** [comment]
Closing this issue. Part of it was fixed, part of it timed out and wasn't reproducable.

**Comment by TorCapybara** [comment]
Only got the Mail from Oct. 25 - Thanks for looking into it. You seem to think that the network is not announcing the Modes "correctly" - I will also look into that one when trying this out the next time.
